### ISSUE TITLE
`shadow_unrelated` should permit `for` loops

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#11827]
### Summary

In any situation where you have a iterable that is tree-like or otherwise subdividable into more of itself — or where the English noun being used for the type does not have distinct singular and plural forms — it may be appropriate to shadow a variable by means of a `for` loop.

I propose that `shadow_unrelated` should apply the same relatedness rules to `for PAT in EXPR` that it would for `let PAT = EXPR`.

### Lint Name

shadow_unrelated

### Reproducer

I tried this code:

```rust
#![deny(clippy::shadow_unrelated)]

fn subdivide(chunk: &[u8]) {
    for chunk in chunk.chunks(10) {
        let _ = chunk;
    }
}
```

I saw this happen:

```
error: `chunk` shadows a previous, unrelated binding
 --> src/lib.rs:4:9
  |
4 |     for chunk in chunk.chunks(10) {
  |         ^^^^^
  |
note: previous binding is here
 --> src/lib.rs:3:13
  |
3 | fn subdivde(chunk: &[u8]) {
  |             ^^^^^
```

I expected to see this happen: No warning

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-apple-darwin
release: 1.74.0
LLVM version: 17.0.4
```
